File: frontend/src/api/modules/nezha-agent.ts
import http from '@/api/http'
import type { NezhaAgentStatus } from '@/views/nezha-agent/state'

export type { NezhaAgentStatus }

/** Allowed lifecycle operations for the bundled Nezha Agent. */
export type NezhaAgentOperation = 'start' | 'stop' | 'restart' | 'enable' | 'disable'

/**
 * Config write payload.
 * clientSecret is optional: omit the field entirely when the user leaves it blank
 * so the backend leaves the existing secret unchanged.
 */
export interface NezhaAgentConfigPayload {
  dashboardUrl: string
  clientSecret?: string
  remoteOperationsEnabled: boolean
  enableAndStart: boolean
}

export const getNezhaAgentStatus = () => {
  return http.get('/nezha-agent/status')
}

export const updateNezhaAgentConfig = (data: NezhaAgentConfigPayload) => {
  return http.put('/nezha-agent/config', data)
}

/** Restore the version-matched bundled binary and systemd unit. No secret body. */
export const installNezhaAgent = () => {
  return http.post('/nezha-agent/install')
}

export const operateNezhaAgent = (operation: NezhaAgentOperation) => {
  return http.post('/nezha-agent/operate', { operation })
}

File: backend/app/api/v1/nezha_agent.go
package v1

import (
	"xpanel/app/api/v1/helper"
	"xpanel/app/dto"
	"xpanel/app/service"

	"github.com/gin-gonic/gin"
)

// NezhaAgentAPI exposes the minimal private panel API for the bundled Nezha Agent.
// Logs are intentionally not exposed here; use /api/v1/toolbox/services/logs
// with the fixed xpanel-nezha-agent unit name.
type NezhaAgentAPI struct{}

// nezhaAgentService is the handler-facing subset of NezhaAgentService.
// Kept minimal so tests can inject fakes without the full service package surface.
type nezhaAgentService interface {
	Status() (*dto.NezhaAgentStatus, error)
	Configure(req dto.NezhaAgentConfigUpdate) error
	Operate(operation string) error
}

// newNezhaAgentService is a package-level factory for handler tests.
// Production defaults to service.NewNezhaAgentService. Tests must restore via cleanup.
var newNezhaAgentService = func() nezhaAgentService {
	return service.NewNezhaAgentService()
}

// GetNezhaAgentStatus returns component/config/service status without secrets.
func (a *NezhaAgentAPI) GetNezhaAgentStatus(c *gin.Context) {
	st, err := newNezhaAgentService().Status()
	if err != nil {
		helper.HandleError(c, err)
		return
	}
	helper.SuccessWithData(c, st)
}

// UpdateNezhaAgentConfig applies a partial config update. Success has no data body.
// clientSecret is accepted for Configure but never logged or echoed by this handler.
func (a *NezhaAgentAPI) UpdateNezhaAgentConfig(c *gin.Context) {
	var req dto.NezhaAgentConfigUpdate
	if err := helper.CheckBindAndValidate(&req, c); err != nil {
		helper.HandleError(c, err)
		return
	}
	if err := newNezhaAgentService().Configure(req); err != nil {
		helper.HandleError(c, err)
		return
	}
	helper.SuccessWithOutData(c)
}

// OperateNezhaAgent runs start|stop|restart|enable|disable. Success has no data body.
func (a *NezhaAgentAPI) OperateNezhaAgent(c *gin.Context) {
	var req dto.NezhaAgentOperateRequest
	if err := helper.CheckBindAndValidate(&req, c); err != nil {
		helper.HandleError(c, err)
		return
	}
	if err := newNezhaAgentService().Operate(req.Operation); err != nil {
		helper.HandleError(c, err)
		return
	}
	helper.SuccessWithOutData(c)
}
